engine/access/index/engine.go

```diff
@@ -0,0 +1,293 @@
+		lg.Fatal().Err(err).Msg("failed to get header for execution data")
+		return
+	}
+
+	if ok := e.highestHeight.Set(header.Height); !ok {
+		// this means that the height was lower than the current highest height
+
+		// OnExecutionData is guaranteed by the requester to be called in order, but may be called
+		// multiple times for the same block.
+		lg.Debug().Msg("execution data for block already received")
+		return
+	}
+
+	e.executionDataNotifier.Notify()
+}
+
+func (e *Engine) processExecutionData(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+	notifier := e.executionDataNotifier.Channel()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-notifier:
+			err := e.processAvailableExecutionData(ctx)
+			if err != nil {
+				// if an error reaches this point, it is unexpected
+				ctx.Throw(err)
+				return
+			}
+		}
+	}
+}
+
+func (e *Engine) processAvailableExecutionData(ctx context.Context) error {
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+
+		// TODO: loop over all heights between our latest and the highest seen, and index
+		// maybe use a jobqueue?
+
+		height := e.lastFullyProcessedHeight.Value() + 1
+
+		// reached the end
+		if height > e.highestHeight.Value() {
+			return nil
+		}
+
+		execData, err := e.getExecutionData(ctx, height)
+		if err != nil {
+			return fmt.Errorf("could not get execution data for block height %d: %w", height, err)
+		}
+
+		if err := e.handleExecutionData(execData); err != nil {
+			return err
+		}
+
+		// TODO: persist the last processed height
+		if ok := e.lastFullyProcessedHeight.Set(height); !ok {
+			// TODO: should this cause the node to crash?
+			return fmt.Errorf("could not set last processed height to %d", height)
+		}
+	}
+}
+
+// getExecutionData returns the execution data for the given block height.
+// Expected errors during normal operation:
+// - storage.ErrNotFound or execution_data.BlobNotFoundError: execution data for the given block height is not available.
+func (e *Engine) getExecutionData(ctx context.Context, height uint64) (*execution_data.BlockExecutionDataEntity, error) {
+	// fail early if no notification has been received for the given block height.
+	// note: it's possible for the data to exist in the data store before the notification is
+	// received. this ensures a consistent view is available to all callers.
+	if height > e.highestHeight.Value() {
+		return nil, fmt.Errorf("execution data for block %d is not available yet: %w", height, storage.ErrNotFound)
+	}
+
+	execData, err := e.execDataCache.ByHeight(ctx, height)
+	if err != nil {
+		return nil, fmt.Errorf("could not get execution data for block %d: %w", height, err)
+	}
+
+	return execData, nil
+}
+
+func (e *Engine) handleExecutionData(execData *execution_data.BlockExecutionDataEntity) error {
+	// TODO: should we parallelize indexing collections and events for each chunk?
+	for i, chunkData := range execData.ChunkExecutionDatas {
```

Comment:
note: the last `chunkData` is for the system chunk. this should be handled differently:
* The collection should not be indexed. It is auto generated as needed.
* All events should be indexed as `ServiceEvent` instead of standard user `Event`